### PR TITLE
sql: make table PUBLIC when references created in transaction

### LINKED ISSUE #24626: sql: can't create table, index and foreign key in the same transaction
This is on cockroachdb 2.0.0 on linux 64 bit.

These statements:
```
start transaction;
create table t_parent (id varchar primary key);
create table t_child (id varchar primary key, parent_id varchar);
create index idx_t_child_parent_id on t_child (parent_id);
alter table t_child add constraint fk_t_child_parent_id foreign key (parent_id) references t_parent (id);
commit;
```
Give this error:
```
pq: foreign key requires an existing index on columns ("parent_id")
```
I am not sure if this is an issue or a known limitation. I am working on a java application which uses automatic migration from the create scripts, but they fail when used in a transaction. I hope to avoid having only the foreign keys moved to their own transaction and separating the index code from its corresponding alter code.

### PR DESCRIPTION
**Pull request by vivekmenezes**
Normally a table with references (FK, interleaved) is
placed in the ADD state to allow the referenced tables
to be upgraded with backreferences across the cluster,
before the cluster sees the ADD table in the PUBLIC state.
When all the table references are created in the same
transaction the table can be made PUBLIC immediately
because all the references are still not visible to
the cluster. This is particularly valuable when commands
within a transaction need to use the PUBLIC table and
will normally barf on seeing a table in the ADD state.

This also fixes a problem with column backfill in the
presence of FKs.

fixes #24626

Release note (sql fix): Fix problems with using tables
created in the same transaction when the tables have FOREIGN
KEY or INTERLEAVED references to other tables which have also
been created in the same transaction.

### CODE REVIEW

**Review by madelynnblue — APPROVED** [review]

File: pkg/sql/backfill.go [hunk 1]
```diff
@@ -677,17 +681,46 @@ func runSchemaChangesInTxn(
 func columnBackfillInTxn(
 	ctx context.Context,
 	txn *client.Txn,
+	tc *TableCollection,
 	evalCtx *tree.EvalContext,
 	tableDesc *sqlbase.TableDescriptor,
 ) error {
+	// A column backfill in the ADD state is a noop.
```
Comment (knz) [inline pkg/sql/backfill.go]: Are you sure?

What about:

```sql
BEGIN;
CREATE TABLE blah ...;
INSERT INTO blah VALUES ...;
CREATE INDEX blih ON blah (...);  -- this needs a backfill!
COMMIT;
```

Comment (vivekmenezes) [reply pkg/sql/backfill.go]: If TABLE blah is in the ADD state, the INSERT will fail

Comment (knz) [reply pkg/sql/backfill.go]: oh I see thanks!

**Pull request merged**